pkg/resources/konnectivity/proxyservice.go

```diff
@@ -0,0 +1,80 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package konnectivity
+
+import (
+	"fmt"
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	"k8c.io/kubermatic/v2/pkg/resources"
+	"k8c.io/kubermatic/v2/pkg/resources/nodeportproxy"
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+// ServiceCreator returns function to create konnectivity proxy service.
+func ServiceCreator(exposeStrategy kubermaticv1.ExposeStrategy, externalURL string) reconciling.NamedServiceCreatorGetter {
+	return func() (string, reconciling.ServiceCreator) {
+		return resources.KonnectivityProxyServiceName, func(se *corev1.Service) (*corev1.Service, error) {
+
+			se.Spec.Selector = map[string]string{
+				resources.AppLabelKey: "apiserver", // because konnectivity proxy runs in sidecar in apiserver pod
+			}
+
+			if se.Annotations == nil {
+				se.Annotations = make(map[string]string)
+			}
+
+			switch exposeStrategy {
+			case kubermaticv1.ExposeStrategyNodePort:
+				se.Spec.Type = corev1.ServiceTypeNodePort
+				se.Annotations[nodeportproxy.DefaultExposeAnnotationKey] = nodeportproxy.NodePortType.String()
+				delete(se.Annotations, nodeportproxy.NodePortProxyExposeNamespacedAnnotationKey)
+			case kubermaticv1.ExposeStrategyLoadBalancer:
+				se.Spec.Type = corev1.ServiceTypeNodePort
+				se.Annotations[nodeportproxy.NodePortProxyExposeNamespacedAnnotationKey] = "true"
+				delete(se.Annotations, nodeportproxy.DefaultExposeAnnotationKey)
+			case kubermaticv1.ExposeStrategyTunneling:
+				se.Spec.Type = corev1.ServiceTypeClusterIP
+				se.Annotations[nodeportproxy.DefaultExposeAnnotationKey] = strings.Join([]string{nodeportproxy.SNIType.String(), nodeportproxy.TunnelingType.String()}, ",")
+				se.Annotations[nodeportproxy.PortHostMappingAnnotationKey] = fmt.Sprintf(`{"secure": %q}`, "konnectivity-server."+externalURL)
+				delete(se.Annotations, nodeportproxy.NodePortProxyExposeNamespacedAnnotationKey)
+			default:
+				return nil, fmt.Errorf("unsupported expose strategy: %q", exposeStrategy)
+			}
+
+			if len(se.Spec.Ports) == 0 {
+				se.Spec.Ports = make([]corev1.ServicePort, 1)
+			}
+
+			const port = 8132
+
+			se.Spec.Ports[0].Name = "secure"
+			se.Spec.Ports[0].Port = 443
+			se.Spec.Ports[0].Protocol = corev1.ProtocolTCP
+			se.Spec.Ports[0].TargetPort = intstr.FromInt(port)
+
+			if exposeStrategy == kubermaticv1.ExposeStrategyTunneling {
+				se.Spec.Ports[0].NodePort = 0 // TODO(pratik): why?
```

Comment:
allows switching from other expose strategies.. which is not possible ATM anyway, but I would still prefer to keep it for consistency with the rest of the code

Reply:
I want to know how exactly that switching would work tho.